yetanotherco/aligned_layer
# fix: use bls signature in telemetry api for the operator

## Description

When sending the operator version to the telemetry service we verify the operator signature using the `ECDSA` key. This pr changes it to verify the signature with the operator `BLS` corresponding to the one registered on the aligned avs.

Something to note is that in the `telemetry` message from the operator we provide the `signature` and the `publickey` in the `G2` generator. The last part is needed as the `BLSAPKRegistry` stores the operator `publickey` on `G1` and the signature is also in `G1`. So the verification has two steps:
1. That the operator provided `G2` `publickey` corresponds to its `G1` equivalence registered on ethereum (we perform the pairing check)
2. The actual signature verification using the `G2` pubkey.

Finally, the code might not be so elixir idiomatic (I am open to any criticism to improve it if necessary).

> Note: when building telemetry now we need to run the following:
> ```shell
> make telemetry_compile_bls_verifier
> ```
> This is already done on `make telemetry_full_start`



**Testing**

You can test it by registering an operator, verifying the signature is well verified, and then changing its config file to point to another `bls` keystore, this would fail as the `G1` -> `G2` correspondence of the keys won't match:
1. Start telemetry: `make telemetry_full_start`
2. Start Aggregator: `make aggregator_start`
3. Register operator: `make operator_full_registration CONFIG_FILE=config-files/config-operator-1.yaml`
4. Build operator: `make build_operator_macos`
5. Start operator: `operator/build/aligned-operator start --config config-files/config-operator-1.yaml`
6. Change line `19` in `config-files/config-operator-1.yaml` for the operator 2 bls key: 
```yaml
  private_key_store_path: "config-files/devnet/keys/operator-2.bls.key.json"
```
7. Re start operator (you should get a `401` err): `operator/build/aligned-operator start --config config-files/config-operator-1.yaml`
8. Go to `http://localhost:4001/api/operators` and check telemetry still works.

## Type of change


## Checklist

  - [Link to the open issue addressing it]()
